From 6b0b443356dd2b44dc5015707898f5faaa51fc1d Mon Sep 17 00:00:00 2001
From: Tristan Cacqueray <tdecacqu@redhat.com>
Date: Sun, 14 May 2017 01:12:17 +0000
Subject: [PATCH] launcher: add results_dir option

This change adds a new results_dir option to instruct the launcher
to store console log like Jenkins. This provides a convenient
overview of executed jobs, in particular for post and periodic pipelines.

Change-Id: I2a7456918e179a9eb1cd8b136cc0b3bfdd84654e
---

diff --git a/zuul/launcher/ansiblelaunchserver.py b/zuul/launcher/ansiblelaunchserver.py
index 834ed62..27bb87b 100644
--- a/zuul/launcher/ansiblelaunchserver.py
+++ b/zuul/launcher/ansiblelaunchserver.py
@@ -622,6 +622,10 @@
         self.library_dir = library_dir
         self.pre_post_library_dir = pre_post_library_dir
         self.options = options
+        if self.config.has_option('launcher', 'results_dir'):
+            self.results_dir = config.get('launcher', 'results_dir')
+        else:
+            self.results_dir = None
 
     def isAlive(self):
         # Meant to be called from the manager
@@ -1378,6 +1382,43 @@
                         when='console_pid_file.stat.exists')
             tasks.append(task)
 
+            if self.results_dir:
+                result_dir = "%s/%s-%s" % (
+                    self.results_dir,
+                    job_name,
+                    parameters.get("ZUUL_PIPELINE"),
+                )
+                job_dir = "%s/%s-%s,%s-%s" % (
+                    result_dir,
+                    parameters.get("ZUUL_PROJECT"),
+                    parameters.get("ZUUL_CHANGE", "0"),
+                    parameters.get("ZUUL_PATCHSET", "0"),
+                    parameters.get("ZUUL_UUID"),
+                )
+                task = dict(file=dict(path="%s-SUCCESS" % job_dir,
+                                      state="directory"),
+                            when='success|bool')
+                tasks.append(task)
+                task = dict(copy=dict(src=console_path,
+                                      dest="%s-SUCCESS/console.html" % job_dir,
+                                      remote_src=True),
+                            when='success|bool')
+                tasks.append(task)
+                task = dict(file=dict(path="%s/lastSuccessful" % result_dir,
+                                      src="%s-SUCCESS/console.html" % job_dir,
+                                      state="link"),
+                            when='success|bool')
+                tasks.append(task)
+                task = dict(file=dict(path="%s-FAILURE" % job_dir,
+                                      state="directory"),
+                            when='not success|bool')
+                tasks.append(task)
+                task = dict(copy=dict(src=console_path,
+                                      dest="%s-FAILURE/console.html" % job_dir,
+                                      remote_src=True),
+                            when='not success|bool')
+                tasks.append(task)
+
             play = dict(hosts='node', name='Publishers',
                         tasks=tasks)
             playbook.write(yaml.safe_dump([play], default_flow_style=False))
